### PR TITLE
PYTHON-3965 Update Langchain vectorSearch with $vectorSearch support

### PR DESCRIPTION
**Pull request by NoahStapp**
[PYTHON-3965](https://jira.mongodb.org/browse/PYTHON-3965)

### CODE REVIEW

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py [hunk 1]
```diff
@@ -159,24 +198,70 @@ def _similarity_search_with_score(
...
+    ) -> List[Tuple[Document, float]]:
+        global _use_vectorsearch
+        if _use_vectorsearch is None:
+            try:
+                result = self._similarity_search_query_vectorsearch(
+                    embedding, k, pre_filter, post_filter_pipeline
+                )
+                _use_vectorsearch = True
+            except OperationFailure as e:
+                if "$vectorSearch is not allowed with the current configuration" in str(e):
```
Comment (ShaneHarvey) [inline libs/langchain/langchain/vectorstores/mongodb_atlas.py]: `"$vectorSearch is not allowed with the current configuration" in str(e):` is very brittle since the server can change the error message at any point. Instead is it possible to rely on the error code?

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py [hunk 2]
```diff
@@ -29,6 +30,7 @@
 logger = logging.getLogger(__name__)
 
 DEFAULT_INSERT_BATCH_SIZE = 100
+_use_vectorsearch = True
```
Comment (ShaneHarvey) [inline libs/langchain/langchain/vectorstores/mongodb_atlas.py]: Is this supposed to start as None?

Edit: this is irrelevant after https://github.com/NoahStapp/langchain/pull/1#discussion_r1332148892

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py [hunk 3]
```diff
@@ -89,6 +91,18 @@ def from_connection_string(
         embedding: Embeddings,
         **kwargs: Any,
     ) -> MongoDBAtlasVectorSearch:
+        """Construct a `MongoDB Atlas Vector Search` vector store
+        from a MongoDB connection URI.
+
+        Args:
+            connection_string: A valid MongoDB connection URI.
```
Comment (ShaneHarvey) [inline libs/langchain/langchain/vectorstores/mongodb_atlas.py:98]: What happens if an app connects to 2 MongoDB clusters, one which supports $vectorSearch and another that does not?

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py [hunk 4]
```diff
@@ -159,24 +198,70 @@ def _similarity_search_with_score(
...
+                        f"$vectorSearch not supported for this Atlas version. "
+                        f"Attempting to use $search. Original error:\n\t{e}"
+                    )
+                    _use_vectorsearch = False
+                    result = self._similarity_search_query_search(
+                        embedding, k, pre_filter, post_filter_pipeline
+                    )
+                else:
+                    raise
+        elif _use_vectorsearch:
```
Comment (ShaneHarvey) [inline libs/langchain/langchain/vectorstores/mongodb_atlas.py]: Can we simplify this into just 2 cases True/False rather than None/True/False? That way we can avoid the duplicated calls to _similarity_search_query_vectorsearch.

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py [hunk 5]
```diff
@@ -254,33 +339,33 @@ def similarity_search(
     def max_marginal_relevance_search(
         self,
         query: str,
-        k: int = 4,
+        limit: int = 4,
```
Comment (ShaneHarvey) [inline libs/langchain/langchain/vectorstores/mongodb_atlas.py]: This seems like an unrelated api change that would be better in a separate PR.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py (hunk 3 above)
Comment (NoahStapp) [reply libs/langchain/langchain/vectorstores/mongodb_atlas.py:98]: Can a query be run on two clusters in a single command? If the query is sent to each cluster separately, moving `_use_vectorsearch` to be an attribute of a MongoDBAtlasVectorSearch should solve that issue.

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py (hunk 4 above)
Comment (NoahStapp) [reply libs/langchain/langchain/vectorstores/mongodb_atlas.py]: We need to have some kind of check if this is the first call or not, since the first call needs to try `$vectorSearch` first and fallback to `$search` if it's unsupported.

Comment (ShaneHarvey) [reply libs/langchain/langchain/vectorstores/mongodb_atlas.py]: Can't every _use_vectorsearch=True call perform the fallback logic?

Comment (NoahStapp) [reply libs/langchain/langchain/vectorstores/mongodb_atlas.py]: Oh right, for some reason I thought that meant repeated attempts to use `$vectorSearch`. Simplified down!

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py (hunk 3 above)
Comment (ShaneHarvey) [reply libs/langchain/langchain/vectorstores/mongodb_atlas.py:98]: > Can a query be run on two clusters in a single command?

No.

> moving _use_vectorsearch to be an attribute of a MongoDBAtlasVectorSearch should solve that issue.

Agreed that seems like the best approach.

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py (hunk 5 above)
Comment (NoahStapp) [reply libs/langchain/langchain/vectorstores/mongodb_atlas.py]: Good catch, thought this was a non-public API for some reason.

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py [hunk 6]
```diff
@@ -159,24 +198,64 @@ def _similarity_search_with_score(
...
+        pre_filter: Optional[dict] = None,
+        post_filter_pipeline: Optional[List[Dict]] = None,
+    ) -> List[Tuple[Document, float]]:
+        if self._use_vectorsearch:
+            try:
+                result = self._similarity_search_query_vectorsearch(
+                    embedding, k, pre_filter, post_filter_pipeline
+                )
+            except OperationFailure as e:
+                if e.code == 224:
```
Comment (ShaneHarvey) [inline libs/langchain/langchain/vectorstores/mongodb_atlas.py]: Could you add a comment here like "QueryFeatureNotAllowed error code"? Found here: https://github.com/mongodb/mongo/blob/0744ea1/src/mongo/base/error_codes.yml#L269C24-L269C46

Could you also confirm that this error code is stable for this use case?

**Review by ShaneHarvey — APPROVED** [review]
LGTM assuming the error code is stable.

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py (hunk 6 above)
Comment (NoahStapp) [reply libs/langchain/langchain/vectorstores/mongodb_atlas.py]: Will do, thanks!

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py [hunk 7]
```diff
@@ -159,24 +198,64 @@ def _similarity_search_with_score(
...
+        self,
+        embedding: List[float],
+        k: int = 4,
+        pre_filter: Optional[dict] = None,
+        post_filter_pipeline: Optional[List[Dict]] = None,
+    ) -> List[Tuple[Document, float]]:
+        params = {
+            "queryVector": embedding,
+            "path": self._embedding_key,
+            "numCandidates": 100,
```
Comment (Jibola) [inline libs/langchain/langchain/vectorstores/mongodb_atlas.py]: I saw in the docs we used 100 and I was a little unclear on the purpose of this key in those docs. Any notion from the vector search devs on how much a developer would need this param tuned?

Comment (NoahStapp) [reply libs/langchain/langchain/vectorstores/mongodb_atlas.py]: Whoops this is supposed to be `k`, not 100. From my understanding having both `limit` and `numCandidates` be `k` is the equivalent of what `$search` does.

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py (hunk 6 above)
Comment (NoahStapp) [reply libs/langchain/langchain/vectorstores/mongodb_atlas.py]: It appears the error code is specific to 6.0.10. Older server versions could give a variety of errors.

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py [hunk 8]
```diff
@@ -60,6 +61,7 @@ def __init__(
         index_name: str = "default",
         text_key: str = "text",
         embedding_key: str = "embedding",
+        compatibility_mode: bool = False,
```
Comment (ShaneHarvey) [inline libs/langchain/langchain/vectorstores/mongodb_atlas.py]: My gut says we should remove this for now. I think the risk a user encounters an issue where they would need to use this feature is low. We can always add this later if we find out it's necessary for some use-case we haven't considered.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py (hunk 8 above)
Comment (NoahStapp) [reply libs/langchain/langchain/vectorstores/mongodb_atlas.py]: Done!

**Review by ShaneHarvey — APPROVED** [review]
